src/overlays/actors/ovl_En_Stone_heishi/z_en_stone_heishi.c

```diff
@@ -35,54 +39,473 @@ const ActorInit En_Stone_heishi_InitVars = {
+    this->action = EN_STONE_ACTION_1;
+    this->actionFunc = func_80BC9560;
+}
+
+void func_80BC9560(EnStoneheishi* this, PlayState* play) {
+    Player* player = GET_PLAYER(play);
+    s16 yawDiff;
+    s32 absYawDiff;
+
+    if (Actor_ProcessTalkRequest(&this->actor, &play->state)) {
+        func_80BC9660(this);
+        return;
+    }
+
+    if ((!(gSaveContext.save.weekEventReg[0x29] & 0x40)) && (play->actorCtx.unk4 != 100)) {
+        this->actor.flags |= ACTOR_FLAG_8000000;
+        return;
+    }
+
+    SkelAnime_Update(&this->skelAnime);
+
+    this->actor.flags &= ~ACTOR_FLAG_8000000;
+    yawDiff = this->actor.yawTowardsPlayer - this->actor.world.rot.y;
+
+    absYawDiff = ABS_ALT(yawDiff);
+
+    if ((absYawDiff < 0x18F1) && ((s32)(player->stateFlags1 << 8) >= 0)) {
+        func_800B8614(&this->actor, play, 70.0f);
+    }
+}
+
+void func_80BC9660(EnStoneheishi* this) {
+    this->textIdSet = false;
+    this->action = EN_STONE_ACTION_2;
+    this->actionFunc = func_80BC9680;
+}
+
+void func_80BC9680(EnStoneheishi* this, PlayState* play) {
+    f32 currentFrame = this->skelAnime.curFrame;
+
+    if ((this->textIndex == 0) || (this->textIndex == 2)) {
+        if (this->animIndex != EN_STONE_HEISHI_SIT_AND_REACH) {
+            if (fabsf(this->headRotY - this->targetHeadPosY) < 50.0f) {
+                EnStoneheishi_ChangeAnim(this, EN_STONE_HEISHI_SIT_AND_REACH);
+            }
+            return;
+        }
+    } else if (this->textIndex == 3) {
+        if (this->animIndex != EN_STONE_HEISHI_WAVE) {
+            if ((this->timer == 0) && (fabsf((f32)(this->headRotY - this->targetHeadPosY)) < 50.0f)) {
+                EnStoneheishi_ChangeAnim(this, EN_STONE_HEISHI_WAVE);
+            }
+            return;
+        } else if (!(this->textIdSet) && (this->endFrame <= currentFrame)) {
+            Player* player = GET_PLAYER(play);
+
+            this->textIdSet = true;
+            player->actor.textId = sEnStoneHeishiTextIds[this->textIndex];
+
+            func_80151938(play, sEnStoneHeishiTextIds[this->textIndex]);
+            func_80151BB4(play, 0x12);
+        }
+    }
+
+    SkelAnime_Update(&this->skelAnime);
+
+    if ((Message_GetState(&play->msgCtx) == TEXT_STATE_5) && Message_ShouldAdvance(play)) {
+        func_801477B4(play);
+
+        if (this->textIndex == 2) {
+            func_80151938(play, 0xFF);
+            EnStoneheishi_SetupCheckGivenItem(this);
+            return;
+        }
+        if (this->textIndex == 3) {
+            func_80BC94B0(this);
+            return;
+        }
+        if (this->textIndex == 6) {
+            EnStoneheishi_GiveItemPrize(this, play);
+            return;
+        }
+        if (this->textIndex < 7) {
+            this->textIndex++;
+        } else {
+            func_80151BB4(play, 0x12);
+            func_80BC94B0(this);
+            return;
+        }
+
+        func_80151938(play, sEnStoneHeishiTextIds[this->textIndex]);
```

Comment:
You should be able to chain this with `if else` to get rid of all the returns. Might need an `else` before the `(this->textIndex < 7)`